src/inflators/text.ts

```diff
@@ -0,0 +1,127 @@
+  opacity?: number;
+  outlineBlur?: number | `${number}%`;
+  outlineColor?: string;
+  outlineOffsetX?: number | `${number}%`;
+  outlineOffsetY?: number | `${number}%`;
+  outlineOpacity?: number;
+  outlineWidth?: number | `${number}%`;
+  overflowWrap?: "normal" | "break-word";
+  sdfGlyphSize?: number;
+  side?: "front" | "back" | "double";
+  strokeColor?: string;
+  strokeOpacity?: number;
+  strokeWidth?: number | `${number}%`;
+  textAlign?: 'left' | 'right' | 'center' | 'justify';
+  textIndent?: number;
+  whiteSpace?: "normal" | "nowrap";
+};
+
+const THREE_SIDES = {
+  front: FrontSide,
+  back: BackSide,
+  double: DoubleSide
+};
+
+// Defaults values must be set for all the optional properties
+// to use foo! operator in the inflator.
+const DEFAULTS: TextParams = {
+  anchorX: "center",
+  anchorY: "middle",
+  clipRect: null,
+  color: "#ffffff",
+  curveRadius: 0,
+  depthOffset: 0,
+  direction: "auto",
+  fillOpacity: 1.0,
+  fontUrl: null,
+  fontSize: 0.075,
+  glyphGeometryDetail: 1.0,
+  gpuAccelerateSDF: true,
+  letterSpacing: 0,
+  lineHeight: "normal",
+  maxWidth: Infinity,
+  opacity: 1.0,
+  outlineBlur: 0,
+  outlineColor: "#000000",
+  outlineOffsetX: 0,
+  outlineOffsetY: 0,
+  outlineOpacity: 1.0,
+  outlineWidth: 0,
+  overflowWrap: "normal",
+  side: "front",
+  strokeColor: "grey",
+  strokeOpacity: 1.0,
+  strokeWidth: 0,
+  textAlign: "center",
+  textIndent: 0,
+  value: "",
+  whiteSpace: "normal"
+};
+
+export function inflateText(world: HubsWorld, eid: number, params: TextParams) {
+  params = Object.assign({}, DEFAULTS, params);
+  const text = new TroikaText();
+  text.material!.toneMapped = false;
+
+  text.text = params.value;
+  text.material!.side = THREE_SIDES[params.side!];
+  text.material!.opacity = params.opacity!;
+  text.font = params.fontUrl!;
+
+  text.anchorX = params.anchorX!;
+  text.anchorY = params.anchorY!;
+  text.clipRect = params.clipRect!;
+  text.color = params.color!;
+  text.curveRadius = params.curveRadius!;
+  text.depthOffset = params.depthOffset!;
+  text.direction = params.direction!;
+  text.fillOpacity = params.fillOpacity!;
+  text.fontSize = params.fontSize!;
+  text.glyphGeometryDetail = params.glyphGeometryDetail!;
+  text.gpuAccelerateSDF = params.gpuAccelerateSDF!;
+  text.letterSpacing = params.letterSpacing!;
+  text.lineHeight = params.lineHeight!;
+  text.maxWidth = params.maxWidth!;
+  text.outlineBlur = params.outlineBlur!;
+  text.outlineColor = params.outlineColor!;
+  text.outlineOffsetX = params.outlineOffsetX!;
+  text.outlineOffsetY = params.outlineOffsetY!;
+  text.outlineOpacity = params.outlineOpacity!;
+  text.outlineWidth = params.outlineWidth!;
+  text.overflowWrap = params.overflowWrap!;
+  text.strokeColor = params.strokeColor!;
+  text.strokeOpacity = params.strokeOpacity!;
+  text.strokeWidth = params.strokeWidth!;
+  text.textAlign = params.textAlign!;
+  text.textIndent = params.textIndent!;
+  text.whiteSpace = params.whiteSpace!;
```

Comment:
As we discussed, we might be able to get rid of the `!`s by defining two types:

`OptionalTextParams` where each of the properties is optional and `TextParams` where each property is required.

The function signature becomes:
```
export function inflateText(world: HubsWorld, eid: number, params: OptionalTextParams) {
```

The defaults become

```
const DEFAULTS: TextParams = {
```

And in the body of the inflator we write:
```
const finalParams : TextParams = Object.assign({}, DEFAULTS, params)
```

Reply:
Unfortunately

```
ERROR in ./src/inflators/text.ts:122:9
TS2322: Type 'TextParams' is not assignable to type 'TextRequiredParams'.
  Types of property 'anchorX' are incompatible.
    Type '"left" | "center" | "right" | undefined' is not assignable to type '"left" | "center" | "right"'.
      Type 'undefined' is not assignable to type '"left" | "center" | "right"'.
    120 | export function inflateText(world: HubsWorld, eid: number, params: TextParams) {
    121 |   params = Object.assign({}, DEFAULTS, params);
  > 122 |   const params2: TextRequiredParams = Object.assign({}, DEFAULTS, params);
        |         ^^^^^^^
    123 |   const text = new TroikaText();
    124 |   text.material!.toneMapped = false;
    125 |
```

Reply:
I experimented. It works (type check passes).

But some disadvantages.

1. Many duplicated type declaration
2. We need to carefully manage that type declarations are synched. Even if types are different between them (ex: string / number) type check didn't catch.

So, I would like to suggest to experiment more and make another PR for it.

Reply:
Typescript has a generic utility type to make all the fields of a type optional (or required) https://www.typescriptlang.org/docs/handbook/utility-types.html#partialtype - Worth looking through the other types in that doc too just to know what tools are available.

So we could either do:
```ts
export function inflateText(world: HubsWorld, eid: number, params: Partial<TextParams>) {
```
or
```ts
const finalParams = Object.assign({}, DEFAULTS, params) as Required<TextParams>;
```

(I think the `Partial` one is more natural)